upbc/protoc-gen-upb.cc

```diff
@@ -369,6 +329,346 @@ std::string CTypeConst(const protobuf::FieldDescriptor* field) {
+  uint64_t GetFieldModifiers(const protobuf::FieldDescriptor* f);
+  void ResolveIntraFileReferences();
+
+  // When we are generating code, tables are linked to sub-tables via name (ie.
+  // a string) rather than by pointer.  We need to emit an initializer like
+  // `&foo_sub_table`.  To do this, we store `const char*` strings in all the
+  // links that would normally be pointers:
+  //    field -> sub-message
+  //    field -> enum table (proto2 only)
+  //    extension -> extendee
+  //
+  // This requires a bit of reinterpret_cast<>(), but it's confined to a few
+  // functions.  We tag the pointer so we know which member of the union to
+  // initialize.
+  enum SubTag {
+    kNull = 0,
+    kMessage = 1,
+    kEnum = 2,
+    kMask = 3,
+  };
+
+  static upb_MiniTable_Sub PackSub(const char* data, SubTag tag);
+  static bool IsNull(upb_MiniTable_Sub sub);
+  void SetSubTableStrings();
+  upb_MiniTable_Sub PackSubForField(const protobuf::FieldDescriptor* f,
+                                    const upb_MiniTable_Field* mt_f);
+  const char* AllocStr(const std::string& str);
+
+ private:
+  typedef absl::flat_hash_map<const protobuf::Descriptor*, upb_MiniTable*>
+      TableMap;
+  typedef absl::flat_hash_map<const protobuf::FieldDescriptor*,
+                              upb_MiniTable_Extension>
+      ExtensionMap;
+  upb::Arena arena_;
+  TableMap table_map_;
+  ExtensionMap extension_map_;
+  upb_MiniTablePlatform platform_;
+};
+
+upb_MiniTable* FilePlatformLayout::GetMiniTable(
+    const protobuf::Descriptor* m) const {
+  auto it = table_map_.find(m);
+  assert(it != table_map_.end());
+  return it->second;
+}
+
+const upb_MiniTable_Extension* FilePlatformLayout::GetExtension(
+    const protobuf::FieldDescriptor* fd) const {
+  auto it = extension_map_.find(fd);
+  assert(it != extension_map_.end());
+  return &it->second;
+}
+
+void FilePlatformLayout::ResolveIntraFileReferences() {
+  // This properly resolves references within a file, in order to set any
+  // necessary flags (eg. is a map).
+  for (const auto& pair : table_map_) {
+    upb_MiniTable* mt = pair.second;
+    // First we properly resolve for defs within the file.
+    for (const auto& f : FieldNumberOrder(pair.first)) {
+      if (f->message_type() && f->message_type()->file() == f->file()) {
+        upb_MiniTable_Field* mt_f = const_cast<upb_MiniTable_Field*>(
+            upb_MiniTable_FindFieldByNumber(mt, f->number()));
+        upb_MiniTable* sub_mt = GetMiniTable(f->message_type());
+        upb_MiniTable_SetSubMessage(mt, mt_f, sub_mt);
+      }
+      // We don't worry about enums here, because resolving an enum will
+      // never alter the mini-table.
+    }
+  }
+}
+
+upb_MiniTable_Sub FilePlatformLayout::PackSub(const char* data, SubTag tag) {
+  uintptr_t val = reinterpret_cast<uintptr_t>(data);
+  assert((val & kMask) == 0);
+  upb_MiniTable_Sub sub;
+  sub.submsg = reinterpret_cast<upb_MiniTable*>(val | tag);
```

Comment:
you need to keep submsg as a `uintptr_t` if you wish to avoid UB here

Reply:
Are you sure? Point 5 here suggests to me that I can count on the pointer to safely round-trip: https://en.cppreference.com/w/cpp/language/reinterpret_cast

It's true that `sub.submsg` has a greater alignment requirement, however I know a priori that the arena aligns all allocations to 16, so the pointer should satisfy the alignment requirement here.